src/libcore/iter.rs

```diff
@@ -369,6 +369,21 @@ pub trait Iterator {
     /// `None` here means that either there is no known upper bound, or the
     /// upper bound is larger than `usize`.
     ///
+    /// # Implementation notes
+    ///
+    /// It is not enforced that an iterator implementation yields
+    /// declared number of elements. A buggy iterator may yield less
+    /// than lower bound or more than upper bound of elements.
+    ///
+    /// `size_hint()` can be used to optimize the fast path, and can be
+    /// used to panic if estimation is incorrect, but must not be used,
```

Comment:
I kinda get what you're saying here, but it's maybe trying to compact too much information. My suggestion:

> size_hint is primarily intended to be used for optimizations such as reserving space for the elements of the iterator. One may assert that the bounds on the size_hint are accurate, but this is not advisable. It is incorrect to assume that size_hint is correct to e.g. omit bounds checks in unsafe code. Implementing size_hint incorrectly should not lead to memory safety violations.